src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -895,6 +895,53 @@ TEST(Aktualizr, InstallWithUpdates) {
   }
 }
 
+/**
+ * Verifies reporting of update download progress
+ *
+ * Checks actions:
+ *
+ * - [x] Report download progress
+ */
+TEST(Aktualizr, ReportDownloadProgress) {
+  // The test initialization part is repeated in many tests so maybe it makes sense
+  // to define a fixture and move this common initialization part into the fixture's SetUp() method
+  TemporaryDirectory temp_dir;
+  auto http = std::make_shared<HttpFake>(temp_dir.Path(), "hasupdates");
+  Config conf = UptaneTestCommon::makeTestConfig(temp_dir, http->tls_server);
+  auto storage = INvStorage::newStorage(conf.storage);
+  Aktualizr aktualizr(conf, storage, http);
+
+  /* perhaps should be static const member of DownloadProgressReport */
+  const unsigned int download_complete_progress_value = {100};
+  unsigned int download_progress = {0};
+  std::function<void(std::shared_ptr<event::BaseEvent> event)> report_event_hdlr =
+      [&](const std::shared_ptr<event::BaseEvent>& event) {
+        /* perhaps should be static const member of DownloadProgressReport */
+        if (event->variant != "DownloadProgressReport") {
+          return;
+        }
+
+        auto download_progress_event = std::dynamic_pointer_cast<event::DownloadProgressReport>(event);
+        ASSERT_NE(download_progress_event, nullptr);
+        ASSERT_NE(download_progress_event.get(), nullptr);
+        EXPECT_GT(download_progress_event->progress, download_progress);
+        download_progress = download_progress_event->progress;
+      };
+
+  aktualizr.SetSignalHandler(report_event_hdlr);
+  aktualizr.Initialize();
+
+  result::UpdateCheck update_result = aktualizr.CheckUpdates().get();
+  ASSERT_EQ(update_result.status, result::UpdateStatus::kUpdatesAvailable);
+  // The test mocks are tailored to emulate a device with a primary ECU and one secondary ECU
+  // for sake of the download progress report testing it's suffice to test it agains just one of the ECUs
+  update_result.updates.pop_back();
+
+  result::Download download_result = aktualizr.Download(update_result.updates).get();
+  EXPECT_EQ(download_result.status, result::DownloadStatus::kSuccess);
+  EXPECT_EQ(download_progress, download_complete_progress_value);
```

Comment:
I think there is one thing missing here: a guarantee that at least two progress reports have actually been seen. As currently written, this test will pass even if none are actually generated, which is what will probably happen if the download is small and fast.

Reply:
No, the test will not pass if none progress reports are generated since download_progress will be equal to zero while download_complete_progress_value is equal to 100.

Reply:
Although, it does not guarantee that at least two reports are received, since just one report with 100 progress value can be sent. The question is if it's an error case ?

Reply:
That's not an error, but it doesn't prove that we are actually sending progress reports if we only get one and it says 100%. We have to see at least two to be able to demonstrate actual progress.